### ISSUE TITLE
Plugin signing + provenance (extend PLUGIN_ALLOW_UNPINNED enforcement)

### ISSUE THREAD

**Issue by HongmingWang-Rabbit** [Molecule-AI/molecule-core#2291]
PLUGIN_ALLOW_UNPINNED gate exists today; extend to require cosign-style signing on plugin distributions. Workspace pulls a plugin → verify signature against operator-trusted key set → refuse on mismatch.

**Comment by HongmingWang-Rabbit** [comment]
Approved as a direction, with the design choices below grounded in how the broader ecosystem is converging.

**Industry context (settles the scheme question):** Sigstore (Fulcio + Rekor) has become the universal provenance layer — npm GA'd `--provenance` in 2023, PyPI shipped digital attestations in late 2024, GitHub Container Registry uses cosign in its default "Publish Docker Container" workflow. Both npm and PyPI now ship "Trusted Publishing" (OIDC-bound publishing from CI, no long-lived signing keys). The convergence is strong enough that picking anything else means leaving the path most of the supply-chain tooling is being built around.

**Pinned design decisions:**

1. **Signing scheme: cosign (sigstore), keyless via Fulcio + Rekor for Molecule-owned plugins.** Short-lived OIDC-bound certs (minutes), no key management. Skip GPG and minisign. Operator-built private plugins keep a keyed-cosign option (operator holds key in their KMS — the AWS KMS we already use).

2. **Trust root: per-operator config in `global_secrets`-style storage** holding (a) a list of trusted Fulcio identities (e.g. `https://github.com/Molecule-AI/*` for keyless) and (b) a list of trusted public keys (for keyed). Both lists support multi-entry from day 1 — no single-key designs. Mirrors how Sigstore policy controllers accept multiple identities.

3. **Modular verification interface.** Plugin loader calls a `Verifier.Verify(artifactDigest) -> VerifyResult` interface. Day-1 impl is `CosignVerifier`; future impls (Trusted-Publishing-only attestations, npm-style provenance for plugins distributed via npm) become a code-only change, not a re-architecture.

4. **Scope: signing only here. SLSA provenance is a separate issue.** Title says "signing + provenance" but they're complementary, not unified. SLSA L3 build attestations (npm's `--provenance`, PyPI's `publish` attestations) require build-system integration and hermeticity — much bigger lift. File a follow-up; cross-link.

5. **Rollout: warn-mode → enforce-mode** (same import-time-hard / runtime-loud pattern as #2298/#2299):
   - Phase 1: verify, log structured warning on mismatch, still load.
   - Phase 2: Molecule-owned plugins all signed; Datadog warn-events at zero → flip `PLUGIN_ALLOW_UNPINNED` default to `false`.
   - Phase 3: keep `PLUGIN_ALLOW_UNPINNED` as the operator escape hatch (incident response).

6. **Signed bundle = artifact + manifest** under one signature. Otherwise a manifest swap redirects a validly-signed binary to a malicious config. Standard practice — cosign signs OCI artifacts including the manifest by default.

7. **Revocation: Rekor entry verification + local denylist.** Rekor gives "was this signature published before key X was revoked"; local `plugin_denylist` table (digest → reason → added_by) covers immediate-block needs.

8. **Verification cache** per workspace: `(plugin_digest, verified_at)`; invalidate on denylist update.

**Fit-to-our-design:** cosign verify is a single static binary call (or Go library import); workspace-server already pulls plugin artifacts at runtime, so adding `Verifier.Verify` in the loader is a contained change. Trust root in `global_secrets` reuses the existing per-operator config path — no new infra. Rekor verification is a network call to a public service, fine from EC2 workspaces; if data-residency requires it later, Sigstore supports private Fulcio/Rekor instances behind the same `Verifier` interface.

**One open question for the design doc:**
- **Offline-verification fallback?** Some operator deployments may run workspaces in air-gapped environments where Rekor isn't reachable. Lean: support a "bundled signature + Rekor inclusion proof" mode (`cosign --bundle`) so verification works offline against pre-fetched proofs. Not phase-1 critical, but design the `Verifier` interface to accommodate.

References for the implementer:
- [Introducing npm package provenance — GitHub Blog](https://github.blog/security/supply-chain-security/introducing-npm-package-provenance/)
- [PyPI now supports digital attestations](https://blog.pypi.org/posts/2024-11-14-pypi-now-supports-digital-attestations/)
- [Safeguard your containers — GitHub Actions container signing](https://github.blog/security/supply-chain-security/safeguard-container-signing-capability-actions/)
- [sigstore/cosign on GitHub](https://github.com/sigstore/cosign)

**Comment by HongmingWang-Rabbit** [comment]
## Phase 1 — current state + design proposal

### Already shipped

- **Pinned-ref enforcement** (`workspace-server/internal/plugins/github.go:71`): production rejects bare `org/repo` specs without `#ref`. `PLUGIN_ALLOW_UNPINNED=true` for local dev only.
- **SHA256 manifest integrity** (`workspace-server/internal/plugins/supply_chain.go`): plugin manifest.json may declare a `sha256` field; staged directory digest is computed and compared. Mismatch fails the install.

### Gap

Pinning + content-hash give **integrity** (the bytes match the SHA the publisher claimed) but not **authenticity** (anyone can create a fork with a matching SHA — there's no proof of WHO produced the artifact). A compromised plugin author or a typo'd `org/repo` (homograph attack — molecule-ai vs molecuie-ai) would slip past today's gates.

### External patterns considered

- **sigstore/cosign** (Kubernetes ecosystem): keyless signing via Fulcio + transparency log via Rekor; verifier asserts the SAN of the signing cert matches a configured identity (e.g. `https://github.com/Molecule-AI/molecule-plugin-foo/.github/workflows/release.yml@refs/tags/v1.0.0`). Adopt: keyless reduces key-management surface; transparency log is auditable.
- **PEP 740 (PyPI provenance attestations)**: PyPI added cosign-based provenance late 2024. Same model — keyless OIDC-from-CI, attested in transparency log. Reject as direct adoption: our plugins aren't on PyPI; they're git refs. But the model itself (keyless OIDC binding to repo+workflow+ref) is what we want.
- **npm provenance** (`--provenance` since npm v9.5): SLSA Level 3-equivalent, also Sigstore-backed. Same shape as PEP 740. Same conclusion.
- **Static-key signing** (release-engineer holds a private key): rejected — key rotation is operationally expensive, key-loss = ecosystem-wide signature replacement. Keyless is strictly better unless air-gapped.

### Proposed design (keyless, OIDC-bound)

1. **Plugin publish workflow** (in each plugin repo):
    ```yaml
    permissions:
      id-token: write   # for OIDC
      contents: read
    - uses: sigstore/cosign-installer@v3
    - run: cosign sign-blob --yes --bundle plugin.bundle.json plugin-archive.tar.gz
    - run: gh release upload v1.0.0 plugin.bundle.json plugin-archive.tar.gz
    ```
    The bundle includes the cert (with SAN tied to the workflow OIDC identity) + signature + Rekor inclusion proof.

2. **Workspace-server resolver** (extend `internal/plugins/`):
    - New: `verifyPluginSignature(stagedDir string, expectedIdentity Identity) error`. Calls `cosign verify-blob` (via Go SDK `sigstore/cosign/v2/cmd/cosign/cli/verify`) against the bundle.
    - `Identity` is `{Issuer: "https://token.actions.githubusercontent.com", SAN: "https://github.com/<org>/<repo>/.github/workflows/<workflow>@refs/tags/<tag>"}` derived from the manifest's `provenance` block.
    - New env: `PLUGIN_REQUIRE_SIGNATURE=true` enforces. Default `false` during rollout (warn-only logging). Flip to default-true once 5 first-party plugins ship signed.

3. **Trust policy** (operator-level):
    - `MOLECULE_TRUSTED_PLUGIN_ISSUERS=https://token.actions.githubusercontent.com` (the only OIDC issuer accepted out of the box).
    - `MOLECULE_TRUSTED_PLUGIN_REPOS=Molecule-AI/*,user-contributed-foo/*` allowlist of repo patterns whose plugins this workspace will accept.
    - Each tenant can override via per-workspace config; per-org override via CP admin.

### SSOT decision

- The cosign verifier lives in `workspace-server/internal/plugins/signature.go` (single Go implementation).
- The Python runtime never re-verifies; it trusts that the workspace-server already verified at install time. (If we ever support direct-from-Python plugin install, that path also calls into Go via subprocess — same SSOT.)
- The trust policy schema lives in `workspace-server/internal/plugins/trust_policy.go` and is the only source for "is this plugin author trusted by this workspace".

### Smallest change first

Stage 1 — keyless verify reader, default OFF:
1. Add cosign verifier code path
2. Add `PLUGIN_REQUIRE_SIGNATURE=false|warn|enforce` env (default `warn`)
3. `warn` mode logs the verification result but always installs

Stage 2 — sign first-party plugins:
4. Add cosign-sign step to molecule-ci's reusable plugin-publish workflow
5. Sign 1 plugin (smallest first-party — pick from #96 sprawl list) via the new flow; observe Rekor entry
6. Verify the workspace-server warn-mode log shows ✅ for the signed plugin and ⚠ for unsigned ones

Stage 3 — fan out:
7. Sign remaining ~12 first-party plugins
8. Flip `PLUGIN_REQUIRE_SIGNATURE` default to `enforce` once all are signed
9. Public docs explain how a community plugin author signs theirs (workflow snippet)

### Security review

| Question | Answer |
|---|---|
| Untrusted input | YES — plugin manifest is attacker-controlled. Signature verification is the protection; cosign-installer is the well-maintained dep boundary. |
| Auth/authz | YES — verification IS the auth check for plugin authors. Failed verification = refuse install. |
| Trust boundary | YES — workspace → plugin author. Today: trust based on pinned ref + SHA. Post-fix: trust based on cryptographic identity (Fulcio cert chain). |
| Logged data | Plugin name, signer SAN, Rekor URL. No secrets. |
| Permission expansion | New service: workspace-server connects to Rekor (rekor.sigstore.dev) at install time. New egress dependency. Mitigation: cache Rekor transparency-log proofs in the manifest bundle, fall back to network only on cache miss; operator can pin Rekor mirror URL. |

Threat model:
- **Malicious plugin author**: stopped by the trust policy `MOLECULE_TRUSTED_PLUGIN_REPOS` allowlist + signature verification.
- **Compromised plugin repo**: attacker who pushes to the repo can't sign because they don't control the OIDC identity (they'd need to compromise GitHub Actions itself). Pinned-ref + SHA already protects against post-tag rewrites.
- **Compromised CI**: attacker who somehow gets OIDC token signs a bad release. Mitigation: trust policy includes workflow-name match (only `.github/workflows/release.yml@v*` accepted, not arbitrary workflows).
- **Sigstore/Fulcio outage**: network egress dependency. Bundle inclusion-proof caching makes verification offline-capable for previously-installed plugins.
- **Quantum / future crypto break**: ECDSA P-256 today; Sigstore migrates to ML-DSA when standardized. Versioned bundle format handles transition.

### Backwards compatibility

- Existing pinned-ref + SHA gates STAY (defense in depth).
- New signature gate is additive. `warn` mode is a no-op for existing plugin install paths.
- Migration: documented runbook explains how to add `cosign sign-blob` to a plugin's release workflow. ~10 lines of YAML per plugin repo.
- Deprecation: never remove pinned-ref or SHA gates — they remain as a fallback for plugins from publishers who can't sign yet (community fork, private repo without Actions, etc.).

### Open questions

1. Should we reject unsigned plugins outright once the rollout is done, or always allow operator-overridden install (with prominent warning)? Defaulting to "operator can override per-install" — same shape as `npm install --allow-unsigned`.
2. Should we cache Rekor inclusion proofs in our manifest format, or always re-verify against live Rekor? Defaulting to live verify with a 1-hour negative cache; can revisit if Rekor latency becomes a tail-latency problem in the install flow.
3. Skipping for now (parked): cross-language SBOM emission (CycloneDX). #92's scope is signing+provenance only.

🤖 Filed by Claude Code (Phase 1 of the methodology — investigation before code)
